Issue_5 Add support for r10k Puppetfile's opts

alex-harvey-z3q opened the pull request:
Add feature for Librarian to recognise r10k's :commit, :tag, and :branch
as accepted aliases for Librarian's :ref. Librarian will fail if more
than one of these options are specified.

A known issue with this implementation is that, internally, all of these
options are still treated as git refs. In other words, Librarian will
not notice if the user calls what is really a tag a commit and so on.

alex-harvey-z3q commented:
I have raised this PR to start a discussion about the general approach.  I have tested that I can now use `:commit` instead of `:ref` in my Librarian-puppet Puppetfile and that works.  I haven't done any other testing.

rnelson0 commented:
Looks good! It will also need a spec test for when two incompatible options are provided.

alex-harvey-z3q committed:
Issue_5  Add support for r10k Puppetfile's opts

Add feature for Librarian to recognise r10k's :commit, :tag, and :branch
as accepted aliases for Librarian's :ref. Librarian will fail if more
than one of these options are specified.

A known issue with this implementation is that, internally, all of these
options are still treated as git refs. In other words, Librarian will
not notice if the user calls what is really a tag a commit and so on.

alex-harvey-z3q commented:
Ok, test added and I cleaned it up a bit.

alex-harvey-z3q commented on lib/librarian/source/git.rb:
```diff
@@ -23,13 +23,17 @@ class Git
-      attr_accessor :uri, :ref, :sha, :path
-      private :uri=, :ref=, :sha=, :path=
+      attr_accessor :uri, :ref, :branch, :tag, :commit, :sha, :path
+      private :uri=, :ref=, :branch=, :tag=, :commit=, :sha=, :path=
```
I must admit I don't really understand what this line is doing, and I have guessed that I should add `:branch=, :tag=, :commit=` in here.

jyaworski commented:
This looks on the right track to me. Awesome.

alex-harvey-z3q commented:
It's working fine for me, so if no one else has any concerns, feel free to merge.

alex-harvey-z3q commented:
Oh hang on, I should update the docs.  Don't merge until I've done that.

alex-harvey-z3q commented:
Ok, related PR raised https://github.com/voxpupuli/librarian-puppet/pull/9

njam commented:
lgtm
